### PR TITLE
Torodb 384 Check BSON types

### PR DESCRIPTION
**Pull request by pgdoval**

### CODE REVIEW

File: engine/backend/common/src/main/java/com/torodb/backend/AbstractIdentifierConstraints.java [hunk 1]
```diff
@@ -64,8 +64,16 @@ protected AbstractIdentifierConstraints(ImmutableSet<String> restrictedSchemaNam
...
+        .put(FieldType.JAVASCRIPT_WITH_SCOPE, 'w') // javascript [w]ith scope
+        .put(FieldType.MIN_KEY, 'k') //min [k]ey
+        .put(FieldType.MAX_KEY, 'a') //m[a]x key
+        .put(FieldType.UNDEFINED, 'u') //[u]ndefined
+        .put(FieldType.MONGO_REGEX, 'g') //Re[g]Ex
+        .put(FieldType.MONGO_DB_POINTER, 'p') //DB [p]ointer
+        .put(FieldType.DEPRECATED, 'z') //running out of available chars; z means "the last".
         .build());
-
+    //Letters left: f h o v
```
Comment (gortiz) [inline engine/backend/common/src/main/java/com/torodb/backend/AbstractIdentifierConstraints.java]: what does that mean? We did consume all letters but these 4?

File: engine/backend/derby/src/main/java/com/torodb/backend/derby/converters/jooq/Decimal128ValueConverter.java [hunk 2]
```diff
@@ -20,48 +20,62 @@
...
+    implements KvValueConverter<JsonObject, String, KvDecimal128> {
 
-  public static final DataType<BigDecimal> DECIMAL128_TYPE = 
-      new DefaultDataType<BigDecimal>(SQLDialect.DERBY, SQLDataType.NUMERIC, "NUMERIC");
+  private static final long serialVersionUID = 1L;
 
-  public static final DataTypeForKv<KvDecimal128> TYPE = DataTypeForKv.from(DECIMAL128_TYPE,
-      new Decimal128ValueConverter());
+  public static final DataTypeForKv<KvDecimal128> TYPE =
+      DataTypeForKv.from(JsonObjectConverter.TYPE, new Decimal128ValueConverter());
```
Comment (gortiz) [inline engine/backend/derby/src/main/java/com/torodb/backend/derby/converters/jooq/Decimal128ValueConverter.java]: Try to add the new line after _DataTypeForKv.from(_, as it was before

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java [hunk 3]
```diff
@@ -95,14 +95,21 @@ public boolean equals(Object obj) {
...
     return true;
-  
   }
 
   @Override
   public <R, A> R accept(KvValueVisitor<R, A> visitor, A arg) {
     return visitor.visit(this, arg);
   }
 
+  public abstract boolean isNaN();
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java]: Attributes must be declared before methods and constructors

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java [hunk 4]
```diff
@@ -276,18 +273,20 @@ private BigDecimal clampAndRound(final BigDecimal initialValue) {
               "Exponent is out of range for Decimal128 encoding of " + initialValue);
         } else {
           BigInteger multiplier = BIG_INT_TEN.pow(diff);
-          value = new BigDecimal(initialValue.unscaledValue().multiply(multiplier),
-              initialValue.scale() + diff);
+          value =
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java]: This seems ugly, it is not necessary to break the line just after the =. Change it to _value = new BigDecimal(_

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java [hunk 5]
```diff
@@ -276,18 +273,20 @@ private BigDecimal clampAndRound(final BigDecimal initialValue) {
...
         // Increasing a very negative exponent may require decreasing precision, which is rounding
-        // Only round exactly (by removing precision that is all zeroes).  
+        // Only round exactly (by removing precision that is all zeroes).
         // An exception is thrown if the rounding would be inexact:
         int diff = initialValue.scale() + MIN_EXPONENT;
         int undiscardedPrecision = ensureExactRounding(initialValue, diff);
         BigInteger divisor = undiscardedPrecision == 0 ? BIG_INT_ONE : BIG_INT_TEN.pow(diff);
-        value = new BigDecimal(initialValue.unscaledValue().divide(divisor),
-            initialValue.scale() - diff);
+        value =
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvDecimal128.java]: the same

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoDbPointer.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  public KvType getType() {
+    return MongoDbPointerType.INSTANCE;
+  }
+
+  @Override
+  public <R, A> R accept(KvValueVisitor<R, A> visitor, A arg) {
+    return visitor.visit(this, arg);
+  }
+
+  public abstract String getNamespace();
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoDbPointer.java]: Attributes must be declared before methods

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoJavascript.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    private static final long serialVersionUID = -5704502002532582458L;
+
+    @Nonnull
+    @Override
+    public String getValue() {
+      return value;
+    }
+
+    private String value;
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoJavascript.java]: Again, attributes before methods

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoJavascriptWithScope.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  public KvType getType() {
+    return JavascriptWithScopeType.INSTANCE;
+  }
+
+  @Override
+  public <R, A> R accept(KvValueVisitor<R, A> visitor, A arg) {
+    return visitor.visit(this, arg);
+  }
+
+  public abstract String getJs();
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/KvMongoJavascriptWithScope.java]: Nope

File: engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/utils/NonIoByteSource.java [hunk 9]
```diff
@@ -171,8 +165,12 @@ private void readObject(java.io.ObjectInputStream in) throws IOException, ClassN
         throw new IOException("The end of the stream was reach before it was expected");
       }
       if (remaining < 0) {
-        throw new IOException("A byte array of lenght " + size + " was expected, but only " + off
-            + " bytes were found");
+        throw new IOException(
```
Comment (gortiz) [inline engine/kvdocument/core/src/main/java/com/torodb/kvdocument/values/utils/NonIoByteSource.java]: Revert this change to the old style

File: engine/kvdocument/json-converter/src/main/java/com/torodb/kvdocument/conversion/json/MapToKvValueConverter.java [hunk 10]
```diff
@@ -55,18 +55,20 @@ public KvDocument convert(Map<String, Object> source) {
...
   private KvValue<?> convertMap(Map<String, Object> source) {
     if (isSpecialObject(source)) {
       return buildSpecialObject(source);
     }
     LinkedHashMap<String, KvValue<?>> docMap = new LinkedHashMap<>();
-    source.forEach((key, value) -> {
-      String interned = key.intern();
-      docMap.put(interned, convertValue(value));
-    });
+    source.forEach(
```
Comment (gortiz) [inline engine/kvdocument/json-converter/src/main/java/com/torodb/kvdocument/conversion/json/MapToKvValueConverter.java]: Revert that to the old style

File: engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/MongoWpConverter.java [hunk 11]
```diff
@@ -99,10 +96,14 @@ public static KvType translate(BsonType bsonType) throws UnsupportedBsonTypeExce
       case DEPRECATED:
         throw new UnsupportedBsonTypeException(bsonType);
       default:
-        throw new AssertionError("It seems that " + bsonType + " has "
-            + "been added to " + BsonType.class.getCanonicalName()
-            + " but it is not defined how to translate it to "
-            + KvType.class.getCanonicalName());
+        throw new AssertionError(
```
Comment (gortiz) [inline engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/MongoWpConverter.java]: Revert that to the old style

File: engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/ToBsonValueTranslator.java [hunk 12]
```diff
@@ -192,9 +210,12 @@ public BsonBinary visit(KvBinary value, Void arg) {
         subtype = BinarySubtype.USER_DEFINED;
         break;
       default:
-        throw new AssertionError("It is not defined how to translate "
-            + "the binary subtype " + value.getSubtype() + " to"
-            + "MongoDB binaries subtypes");
+        throw new AssertionError(
```
Comment (gortiz) [inline engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/ToBsonValueTranslator.java]: Revert that to the old style

File: engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/UnsupportedBsonTypeException.java [hunk 13]
```diff
@@ -20,23 +20,22 @@
...
 public class UnsupportedBsonTypeException extends RuntimeException {
 
   private static final long serialVersionUID = 1L;
 
   private final BsonType bsonType;
 
   public UnsupportedBsonTypeException(BsonType bsonType) {
-    super(bsonType + " is not supported on KVDocument model and should be "
-        + "consumed by a layer closer to MongoDB");
+    super(
```
Comment (gortiz) [inline engine/kvdocument/mongowp-converter/src/main/java/com/torodb/kvdocument/conversion/mongowp/UnsupportedBsonTypeException.java]: Revert that to the old style

File: engine/mongodb/mongodb-core/src/test/java/com/torodb/mongodb/core/MongoDbBsonTypesTest.java [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    Status<FindCommand.FindResult> findResultStatus =
+            writeTransaction.execute(request, FindCommand.INSTANCE, findArgument);
+
+    writeTransaction.close();
+
+
+    //Asserting that the status of both operations are right
+
+    assertTrue(insertResultStatus.getResult().getWriteErrors().isEmpty());
+    assertTrue(insertResultStatus.isOk()?"ok":insertResultStatus.getErrorMsg() + " in collection " + collName, insertResultStatus.isOk());
```
Comment (gortiz) [inline engine/mongodb/mongodb-core/src/test/java/com/torodb/mongodb/core/MongoDbBsonTypesTest.java]: Why are you using ternary operator (operator ?) ?

File: engine/backend/common/src/main/java/com/torodb/backend/AbstractIdentifierConstraints.java (hunk 1 above)
Comment (pgdoval) [inline engine/backend/common/src/main/java/com/torodb/backend/AbstractIdentifierConstraints.java]: Yes, these are the only letters left. I thought it would be useful for the next time we need to add a field

Comment (gortiz) [inline engine/backend/common/src/main/java/com/torodb/backend/AbstractIdentifierConstraints.java]: thats bad :(

File: engine/mongodb/mongodb-core/src/test/java/com/torodb/mongodb/core/MongoDbBsonTypesTest.java (hunk 14 above)
Comment (pgdoval) [inline engine/mongodb/mongodb-core/src/test/java/com/torodb/mongodb/core/MongoDbBsonTypesTest.java]: Because if it was ok, the getErrorMsg was throwing an Exception. It is ugly but I didn't realize it because it is a message in a test class

Comment (gortiz) [inline engine/mongodb/mongodb-core/src/test/java/com/torodb/mongodb/core/MongoDbBsonTypesTest.java]: in that case, please use a if-then-else with assertFail. It is easier to read

File: engine/backend/derby/src/main/java/com/torodb/backend/derby/converters/jooq/Decimal128ValueConverter.java (hunk 2 above)
Comment (pgdoval) [inline engine/backend/derby/src/main/java/com/torodb/backend/derby/converters/jooq/Decimal128ValueConverter.java]: I can't, 102 characters in line (the first attribute changed and is longer now)
